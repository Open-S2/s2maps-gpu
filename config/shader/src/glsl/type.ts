import { TypeLike } from '../types';

export const toTypeString = (t: any): string => {
  if (typeof t === 'string') return t;
  if (t.name) return t.name;
  return t as any;
}

export const toTypeArgs = (t: any[]): any[] => {
  return t?.map(toTypeString) ?? [];
};
