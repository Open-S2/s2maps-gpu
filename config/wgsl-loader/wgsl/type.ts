import { TypeLike, ParameterLike } from '../types';

export const getTypeName = (s: string) => {
  const i = s.indexOf('<');
  return i >= 0 ? s.slice(0, i) : s;
};

export const getAttributeName = (s: string) => {
  const i = s.indexOf('(');
  return i >= 0 ? s.slice(0, i) : s;
};

export const getAttributeArgs = (s: string) => {
  const i = s.indexOf('(');
  const j = s.lastIndexOf(')');
  if (i >= 0 && j >= 0) return s.slice(i + 1, j);
  return null;
};

export const toTypeString = (t: TypeLike): string => {
  if (typeof t === 'object') return t.name;
  return t;
};

export const toTypeArgs = (t: ParameterLike[]): string[] => {
  return t?.map(p => typeof p === 'object' ? toTypeString(p.type) : p) ?? [];
};
