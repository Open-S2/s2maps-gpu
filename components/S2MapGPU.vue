<template>
  <div id="map" ref='container'>
    <slot />
  </div>
</template>

<script lang="ts">
import type { MapOptions, S2Map } from 's2'

export default {
  name: 'S2MapGPU',
  props: {
    mapOptions: {
      type: Object as PropType<MapOptions>,
      required: true
    },
    mapReady: Function as PropType<(s2map: S2Map) => void>
  },
  setup ({ mapOptions, mapReady }) {
    mapOptions = toRaw(mapOptions)
    const container = ref<HTMLElement>()
    const mapInstance = ref<S2Map | null>(null)

    onMounted(async () => {
      const { S2Map } = await import('s2')
      const config = useRuntimeConfig()
      const options: MapOptions = {
        urlMap: {
          apiURL: config.public.dataURL,
          baseURL: config.public.baseURL
        },
        ...mapOptions,
        container: container.value
        // contextType: 2
      }
      const map = new S2Map(options)

      if (typeof mapReady === 'function') map.addEventListener('ready', () => { mapReady(map) }, { once: true })

      mapInstance.value = map
    })

    onUnmounted(() => {
      if (mapInstance.value !== null) mapInstance.value.delete()
    })

    return { container, mapInstance }
  }
}
</script>

<style>
@import url("../assets/styles/s2maps.css");
#map {
  position: absolute;
  top: 0;
  bottom: 0;
  width: 100%;
}
</style>
