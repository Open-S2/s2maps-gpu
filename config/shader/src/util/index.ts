export * from './bundle';
export * from './tree';
