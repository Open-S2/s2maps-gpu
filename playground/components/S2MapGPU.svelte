<script lang="ts">
  import type { Snippet } from 'svelte';
  import type { MapOptions, S2Map } from 's2';

  // --- Props ---
  const { mapOptions, mapReady, children } = $props<{
    mapOptions: MapOptions;
    mapReady?: (s2map: S2Map) => void;
    children?: Snippet;
  }>();

  // --- Internal State ---
  let container = $state<HTMLElement | undefined>(undefined);
  let mapInstance = $state<S2Map | undefined>(undefined);

  $effect(() => {
    let map: S2Map | undefined = undefined; // Temporary reference for cleanup robustness
    /** Initialize the map */
    async function initializeMap() {
      const { S2Map } = await import('s2');

      const options: MapOptions = {
        urlMap: {
          baseURL: 'http://localhost:3000',
          dataURL: 'http://localhost:3000',
          apiURL: 'http://localhost:3000/api',
        },
        attributionOff: true,
        watermarkOff: true,
        controls: false,
        ...mapOptions,
        container,
      };

      mapInstance = map = new S2Map(options);

      if (typeof mapReady === 'function') {
        map.addEventListener(
          'ready',
          () => {
            mapReady(map!);
          },
          { once: true },
        );
      }
    }

    void initializeMap();

    return () => {
      const mapToDelete = mapInstance ?? map;
      mapToDelete?.delete();
      mapInstance = undefined;
    };
  });
</script>

<div id="map" bind:this={container}>
  {@render children?.()}
</div>

<style>
  @import '../../s2/s2maps.css';
  #map {
    /* Condensed style */
    position: absolute;
    top: 0;
    bottom: 0;
    width: 100%;
    height: 100%;
  }
</style>
