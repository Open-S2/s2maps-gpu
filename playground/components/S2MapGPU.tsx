import React, { useEffect, useRef, useState } from 'react';

import type { MapOptions, S2Map } from 's2/index';

/**
 * Props for the S2Map React Component.
 * Props are:
 * - `mapOptions`: The options for the S2Map instance.
 * - `mapReady?`: A function to be called when the S2Map instance is ready. Optional
 * - `children?`: Child elements to be rendered inside the S2Map instance. Optional
 */
export interface S2MapReactComponentProps {
  mapOptions: MapOptions;
  mapReady?: (s2map: S2Map) => void;
  children?: React.ReactNode;
}

/**
 * S2Map React Component
 * @param props - Props passed to the component
 * @returns - The S2Map React Component
 */
export const S2MapGPU: React.FC<S2MapReactComponentProps> = (props: S2MapReactComponentProps) => {
  const { mapOptions, mapReady, children } = props;
  const containerRef = useRef<HTMLDivElement | null>(null);
  const [mapInstance, setMapInstance] = useState<S2Map | null>(null);

  useEffect(() => {
    let map: S2Map | null = null;

    /** Initialize the map */
    const initializeMap = async () => {
      const { S2Map } = await import('s2');

      if (containerRef.current !== null) {
        const options: MapOptions = {
          urlMap: {
            baseURL: 'http://localhost:3000',
            dataURL: 'http://localhost:3000',
          },
          attributionOff: true,
          watermarkOff: true,
          controls: false,
          ...mapOptions,
          container: containerRef.current,
        };
        map = new S2Map(options);
        setMapInstance(map);

        if (typeof mapReady === 'function') {
          map.addEventListener(
            'ready',
            () => {
              mapReady(map!); // Assert map is not null here as it's just been created
            },
            { once: true },
          );
        }
      }
    };

    void initializeMap();

    return () => {
      if (mapInstance !== null) mapInstance.delete();
    };
  }, [mapOptions, mapReady]);

  return (
    <div
      id="map"
      ref={containerRef}
      style={{ position: 'absolute', top: 0, bottom: 0, width: '100%', height: '100%' }}
    >
      {children}
    </div>
  );
};
