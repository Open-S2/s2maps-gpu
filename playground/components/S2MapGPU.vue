<template>
  <div id="map" ref="container">
    <slot />
  </div>
</template>

<script lang="ts">
import { onMounted, onUnmounted, ref, toRaw } from 'vue';

import type { MapOptions, S2Map } from 's2';
import type { PropType, Ref } from 'vue';

/** The exported container and mapInstance */
export interface S2MapComponent {
  container: Ref<HTMLElement | undefined>;
  mapInstance: Ref<S2Map | null>;
}

export default {
  name: 'S2MapGPU',
  props: {
    mapOptions: {
      type: Object as PropType<MapOptions>,
      required: true,
    },
    mapReady: {
      type: Function as PropType<(s2map: S2Map) => void>,
      required: false,
      default: undefined,
    },
  },
  /**
   * Setup the component, map instance, and mounting/unmounting config
   * @param props - Props passed to the component
   * @returns container and mapInstance
   */
  setup(props): S2MapComponent {
    let { mapOptions, mapReady } = props;
    mapOptions = toRaw(mapOptions);
    const container = ref<HTMLElement>();
    const mapInstance = ref<S2Map | null>(null) as Ref<S2Map | null>;

    onMounted(async () => {
      const { S2Map } = await import('s2');
      const options: MapOptions = {
        urlMap: {
          baseURL: 'http://localhost:3000',
          dataURL: 'http://localhost:3000',
          apiURL: 'http://localhost:3000/api',
        },
        attributionOff: true,
        watermarkOff: true,
        controls: false,
        ...mapOptions,
        container: container.value,
      };
      const map = new S2Map(options);

      /** Used by playwright to ensure the map is ready to render */
      window.testMap = toRaw(map);

      if (typeof mapReady === 'function')
        map.addEventListener(
          'ready',
          () => {
            mapReady(map);
          },
          { once: true },
        );

      mapInstance.value = map;
    });

    onUnmounted(() => {
      if (mapInstance.value !== null) mapInstance.value.delete();
    });

    return { container, mapInstance };
  },
};
</script>

<style>
@import url('../../s2/s2maps.css');
#map {
  position: absolute;
  top: 0;
  bottom: 0;
  width: 100%;
  height: 100%;
}
</style>
