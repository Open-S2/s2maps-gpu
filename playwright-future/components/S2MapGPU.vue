<template>
  <div id='map' ref='container'>
    <slot />
  </div>
</template>

<script lang='ts'>
import type { MapOptions, S2Map as S2MapType } from 's2'

declare global {
  interface Window { testMap: S2MapType }
}

export default {
  name: 'S2MapGPU',
  props: {
    mapOptions: {
      type: Object as PropType<MapOptions>,
      required: true
    }
  },
  setup ({ mapOptions }): { container: Ref<HTMLElement | undefined>, mapInstance: Ref<S2MapType | null> } {
    mapOptions = toRaw(mapOptions)
    const container = ref<HTMLElement>()
    const mapInstance = ref<S2MapType | null>(null) as Ref<S2MapType | null>

    onMounted(async () => {
      const options: MapOptions = {
        urlMap: {
          baseURL: 'http://localhost:8080',
          dataURL: 'http://localhost:8080'
        },
        attributionOff: true,
        watermarkOff: true,
        controls: false,
        ...mapOptions,
        container: container.value,
        // TODO: When flat build can handle offscreen, remove this
        offscreen: false
      }
      const map = new window.S2Map(options)

      /** Used by playwright to ensure the map is ready to render */
      window.testMap = toRaw(map)

      mapInstance.value = map
    })

    onUnmounted(() => {
      if (mapInstance.value !== null) mapInstance.value.delete()
    })

    return { container, mapInstance }
  }
}
</script>

<style>
/* @import url('../../assets/styles/s2maps.css'); */
#map {
  position: absolute;
  top: 0;
  bottom: 0;
  width: 100%;
  height: 100%;
}
</style>
