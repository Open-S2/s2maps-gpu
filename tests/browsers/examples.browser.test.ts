import S2MapGPU from '../../components/S2MapGPUTest.vue';
import { page } from '@vitest/browser/context';
import { render } from 'vitest-browser-vue';
import { waitMap } from './util.js';
import { expect, test } from 'vitest';
// styles
import S2BackgroundStyle from '../../pages/s2/background/style.js';
import S2FillStyle from '../../pages/s2/fill/style.js';

import type { GPUType, StyleDefinition } from 's2/index.js';

/**
 * Render test
 * @param snapshotName - name of the snapshot
 * @param style - S2 Map Style
 * @param contextType - GPU Type (1 = WebGL, 2 = WebGL2, 3 = WebGPU)
 * @returns a Vitest-compatible test function
 */
function testRender(snapshotName: string, style: StyleDefinition, contextType: GPUType) {
  return async () => {
    await page.viewport(1920, 1080);

    render(S2MapGPU, {
      props: { mapOptions: { style, contextType } },
    });

    while (typeof window.testMap === 'undefined') {
      await new Promise((r) => setTimeout(r, 50));
    }

    const success = await waitMap();
    expect(success).toBe(true);

    const screen = await page.screenshot();
    expect(screen).toMatchSnapshot(snapshotName);
  };
}

// WebGL2 //

test('S2->Background->WebGL2', testRender('S2_Background_WebGL2', S2BackgroundStyle, 2));
test('S2->Fill->WebGL2', testRender('S2_Fill_WebGL2', S2FillStyle, 2));
