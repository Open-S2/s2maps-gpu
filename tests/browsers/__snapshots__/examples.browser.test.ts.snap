// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`S2->Background->WebGL2 > S2_Background_WebGL2 1`] = `"/Users/craigoconnor/Documents/Projects/OpenS2/s2maps-gpu/tests/browsers/__screenshots__/examples.browser.test.ts/S2--Background--WebGL2-1.png"`;

exports[`S2->Fill->WebGL2 > S2_Fill_WebGL2 1`] = `"/Users/craigoconnor/Documents/Projects/OpenS2/s2maps-gpu/tests/browsers/__screenshots__/examples.browser.test.ts/S2--Fill--WebGL2-1.png"`;
